lib/charms/operator_libs_linux/v0/passwd.py

```diff
@@ -64,604 +45,208 @@
+    """Check if a user exists.
 
-    @property
-    def message(self):
-        """Return the message passed as an argument."""
-        return self.args[0]
+    Args:
+        user: username or gid of user whose existence to check
 
+    Raises:
+        TypeError: where neither a string or int is passed as the first argument
+    """
+    try:
+        if type(user) is int:
+            return pwd.getpwuid(user)
+        elif type(user) is str:
+            return pwd.getpwnam(user)
+        else:
+            raise TypeError("specified argument '%r' should be a string or int", user)
+    except KeyError:
+        logger.info("specified user '%s' doesn't exist", str(user))
+        return None
 
-class UserState(Enum):
-    """The state of a user on the system or in the cache."""
-
-    Present = "present"
-    Absent = "absent"
-    Disabled = "disabled"
-    NoLogin = "nologin"
 
+def group_exists(group: Union[str, int]) -> Optional[grp.struct_group]:
+    """Check if a group exists.
 
-class UserError(Error):
-    """Raised when there's an error managing a user account."""
+    Args:
+        group: username or gid of user whose existence to check
 
+    Raises:
+        TypeError: where neither a string or int is passed as the first argument
+    """
+    try:
+        if type(group) is int:
+            return grp.getgrgid(group)
+        elif type(group) is str:
+            return grp.getgrnam(group)
+        else:
+            raise TypeError("specified argument '%r' should be a string or int", group)
+    except KeyError:
+        logger.info("specified group '%s' doesn't exist", str(group))
+        return None
+
+
+def add_user(
+    username: str,
+    password: Optional[str] = None,
+    shell: str = "/bin/bash",
+    system_user: bool = False,
+    primary_group: str = None,
+    secondary_groups: List[str] = None,
+    uid: int = None,
+    home_dir: str = None,
+) -> str:
+    """Add a user to the system.
+
+    Will log but otherwise succeed if the user already exists.
+
+    Arguments:
+        username: Username to create
+        password: Password for user; if ``None``, create a system user
+        shell: The default shell for the user
+        system_user: Whether to create a login or system user
+        primary_group: Primary group for user; defaults to username
+        secondary_groups: Optional list of additional groups
+        uid: UID for user being created
+        home_dir: Home directory for user
+
+    Returns:
+        The password database entry struct, as returned by `pwd.getpwnam`
+    """
+    try:
+        user_info = pwd.getpwnam(username)
+        logger.info("user '%s' already exists", username)
+        if uid:
+            user_info = pwd.getpwuid(int(uid))
+            logger.info("user with uid '%d' already exists", uid)
+    except KeyError:
+        logger.info("creating user '%s'", username)
+        cmd = ["useradd", "--shell", shell]
```

Comment:
I think the following is a bit flatter/more Pythonic. (Though I understand if you'd like to avoid returns in the middle of functions.)

```suggestion
    try:
        if uid:
            user_info = pwd.getpwnam(username)
            logger.info("user '%s' already exists", username)
            return user_info
        user_info = pwd.getpwuid(int(uid))
        logger.info("user with uid '%d' already exists", uid)
        return user_info
    except KeyError:
        logger.info("creating user '%s'", username)
        
    cmd = ["useradd", "--shell", shell]
```